docs/technical-design.md

```diff
@@ -0,0 +1,130 @@
+```go
+func Reconcile(ctx context.Context, o reconcile.Request) (reconcile.Result, error) {
+	// Implement business logic of reading and writing objects here
+	return reconcile.Result{RequeueAfter: : 5 * time.Minute}, nil
+}
+```
+
+The biggest challenge in deciding on an appropriate interval is that the time required to perform the reconciliation can vary a lot. Small changes may only require a 
+restart of the sidecar proxies or the Ingress gateway and are therefore much faster than a new installation or a Canary upgrade.  
+Given this dynamic in execution, we must always consider the slowest reconciliation process when defining `SyncPeriod`. Also, we have to consider that if we later add another controller to this operator, this new controller will also use the same frequency.  
+Using `RequeueAfter` on the other hand gives us more freedom, but also makes it a bit more complex to understand what is being done and why.
+
+We have decided to use `RequeueAfter` with a frequency of 5 minutes, as this gives us the ability to perform the matching as often as possible without the risk of repeatedly checking the slowest reconciliation process or piling up requests in the reconciliation queue.
+
+The queuing of reconciliation requests is handled by [controller-runtime](https://pkg.go.dev/sigs.k8s.io/controller-runtime) and is out of scope of this design.
+
+### Components
+#### IstioReconciler
+This is the reconciler that takes care of the entire Istio reconciliation process. 
+The responsibility of this reconciler is to fetch the current configuration and pass it on to the other reconciler together with the desired configuration. 
+It also controls the reconciliation process by running the reconcilers considering the execution dependencies between them.
+This reconciler should be run at least every 20 minutes.
+
+#### IstioInstallationReconciler
+This reconciler decides if an installation, upgrade or uninstall of Istio in the cluster must be done. The reconciler also creates the IstioOperator
+which is used to make changes to the Istio installation by passing it to the `IstioManager`.
+
+The installed IstioOperator is created by merging the `IstioCR` with the IstioOperator with Kyma default values.
+
+##### IstioManager
+This component contains the logic for managing the Istio installation. It knows about the supported client versions and forwards the 
+Istio API requests (e.g. Install, Upgrade) to the correct version of `IstioClient` if there are breaking changes.  
+Depending on how the Istio Go module behaves, it may not be necessary to support multiple clients for the time being. For the first simple implementation, we can probably do without the IstioManager for the time being.
+
+##### IstioClient
+A IstioClient encapsulates a specific version of the [Istio Go module](https://github.com/istio/istio). 
+As we want to support canary updates at some point we might need to support two version of the library if there are breaking changes.
```

Comment:
We added the support for multiple versions because of one disadvantage mentioned in the [evaluation of Kyma Istio Operator options](https://github.tools.sap/xf-goat/kyma-istio-operator). There it is mentioned that there might be [breaking changes to an interface](https://github.tools.sap/xf-goat/kyma-istio-operator/blob/ec0f99786408407b4a6d8b79abe3af6c389cd35d/README.md?plain=1#L22). 
This alone didn't influence design, but there is some uncertainty how the Istio Go module will evolve. 
If the assumption is correct that installing Istio in version 1.13 also works for uninstalling with it with the version 1.14, we do not need to support two versions as mentioned in the [design](https://github.com/triffer/istio/blob/745581abace17663a730efe7f801f12325cb2ed9/docs/technical-design.md?plain=1#L98). But we need to consider that using the API directly is no officially supported way of installing Istio. There might be some implementations in the supported installation scenarios that take care of this. 
Again, this is really hypothetical and we preferred to draw a solution for this, even if hopefully we will never need it. 
The implementation should be modular enough, so we can easily add support for multiple versions if we have the requirement.